File: src/util/int8.fs
module Int8

(* Represent signed chars internally with the int32 type *)
type t = int32

let equal (a: t) (b: t) = a = b
let compare (a: t) (b: t) = compare a b

let show (i8: t) = i8.ToString()

let zero: t = 0

(* internal function to sign-or-zero-extend into upper bytes *)
let resetUpperBytes (x: t) =
    if x &&& 128 = 0 then
        (* result is positive, zero out upper bits *)
        let bitmask = 0x000000ff
        x &&& bitmask
    else
        (* result is negative, set upper bits to 1 *)
        let bitmask = 0xffffff00 |> int32
        x ||| bitmask

let ofInt (i: int) : t = resetUpperBytes (int32 i)
let toInt (x: t) : int = int x
let ofInt64 (i: int64) : t = resetUpperBytes (int32 i)
let toInt64 (x: t) : int64 = int64 x
let toString (x: t) : string = x.ToString()

File: src/util/Int8.rs
//! Signed-byte values represented as `i32`, matching the F# implementation.

pub type Int8Value = i32;

pub fn equal(a: Int8Value, b: Int8Value) -> bool { a == b }
pub fn compare(a: Int8Value, b: Int8Value) -> std::cmp::Ordering { a.cmp(&b) }
pub fn show(value: Int8Value) -> String { value.to_string() }
pub const ZERO: Int8Value = 0;

pub fn reset_upper_bytes(value: Int8Value) -> Int8Value {
    if value & 128 == 0 { value & 0xff } else { value | !0xff }
}

pub fn of_int(value: i64) -> Int8Value { reset_upper_bytes(value as i32) }
pub fn to_int(value: Int8Value) -> i32 { value }
pub fn of_int64(value: i64) -> Int8Value { reset_upper_bytes(value as i32) }
pub fn to_int64(value: Int8Value) -> i64 { i64::from(value) }
pub fn to_string(value: Int8Value) -> String { value.to_string() }

/// Bit-preserving conversion useful when emitting a byte.
pub fn to_uint8(value: Int8Value) -> u8 { value as u8 }
